### PR TITLE
Miscellaneous feature enhancements to BitVector

### PR DESCRIPTION
**Pull request by perryk12**
Address issue #261 by adding PushBack, PopBack, Insert, Delete, MaskHigh, and MaskLow functions.

### CODE REVIEW

**Comment by perryk12** [comment]
Currently the delete function for middle bits is using a naive approach. This should be updated using a Bit Magic strategy similar to Insert.

File: include/emp/bits/BitVector.hpp [hunk 1]
```diff
@@ -701,6 +700,72 @@ namespace emp {
...
+      size_t field_id = FieldID(index);
+      int field_t_bits = sizeof(bit_set[0]) * 8;
+      for (size_t i = field_id + 1; i < NumFields(); i++)
+        bit_set[i] = 0;
+      if (index < num_bits)
+        bit_set[field_id] &= MaskLow<field_t>(index % field_t_bits);
+    }
+
+    /// Insert bits into vector, including middle using bitmagic
+    /// https://devolab.org/?p=2249
```
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:739]: Give a little more explanation of what this link is

File: include/emp/bits/BitVector.hpp [hunk 2]
```diff
@@ -701,6 +700,72 @@ namespace emp {
...
+    void Mask_High(size_t index) {
+      size_t field_id = FieldID(index);
+      int field_t_bits = sizeof(bit_set[0]) * 8;
+      for (size_t i = 0; i < field_id; i++)
+        bit_set[i] = 0;
+      if (index >= 0)
+        bit_set[field_id] &= (MaskLow<field_t>((num_bits - index - 1) % field_t_bits) << (index + 1));
+    }
+
+    // Returns mask that keeps bits lower than index
```
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:728]: explicitly say whether index is included (e.g., lower than but not including or lower than and including)

File: include/emp/bits/BitVector.hpp [hunk 3]
```diff
@@ -701,6 +700,72 @@ namespace emp {
...
+    /// Pop the last bit in the vector
+    bool PopBack() {
+      bool val = Get(num_bits-1);
+      if (val)  Set(num_bits-1, 0);
+      Resize(num_bits - 1);
+      return val;
+    }
+
+    /// Push given bit(s) onto the vector
+    void PushBack(bool bit=true, size_t num=1) {
```
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:712]: for functions that take params, be sure to use @param to document what they are

see https://developer.lsst.io/cpp/api-docs.html

File: include/emp/bits/BitVector.hpp [hunk 4]
```diff
@@ -701,6 +700,72 @@ namespace emp {
       return (int) (pos_found + (field_id * FIELD_BITS));
     }
 
+    /// Pop the last bit in the vector
+    bool PopBack() {
+      bool val = Get(num_bits-1);
+      if (val)  Set(num_bits-1, 0);
```
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:706]: I would remove the if statement. If the value is already zero, we can just set it to zero again no problem

File: include/emp/bits/BitVector.hpp [hunk 5]
```diff
@@ -701,6 +700,72 @@ namespace emp {
       return (int) (pos_found + (field_id * FIELD_BITS));
     }
 
+    /// Pop the last bit in the vector
+    bool PopBack() {
+      bool val = Get(num_bits-1);
```
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:705]: const bool val

File: include/emp/bits/BitVector.hpp [hunk 6]
```diff
@@ -701,6 +700,72 @@ namespace emp {
       return (int) (pos_found + (field_id * FIELD_BITS));
     }
 
+    /// Pop the last bit in the vector
+    bool PopBack() {
```
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:704]: Describe @return in docstring

File: include/emp/bits/BitVector.hpp [hunk 7]
```diff
@@ -701,6 +700,72 @@ namespace emp {
...
+    /// Push given bit(s) onto the vector
+    void PushBack(bool bit=true, size_t num=1) {
+      Resize(num_bits + num);
+      for (size_t i=0; i < num; i++)
+        Set(num_bits-1, bit);
+    }
+
+    // Returns mask that keeps bits higher than index
+    void Mask_High(size_t index) {
+      size_t field_id = FieldID(index);
```
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:720]: const field_id?

File: include/emp/bits/BitVector.hpp (hunk 3 above)
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:712]: mark args in functions const  (ex const bool bit, const size_t num etc ) for all functions

File: include/emp/bits/BitVector.hpp [hunk 8]
```diff
@@ -701,6 +700,72 @@ namespace emp {
...
+      for (size_t i = 0; i < field_id; i++)
+        bit_set[i] = 0;
+      if (index >= 0)
+        bit_set[field_id] &= (MaskLow<field_t>((num_bits - index - 1) % field_t_bits) << (index + 1));
+    }
+
+    // Returns mask that keeps bits lower than index
+    void Mask_Low(size_t index) {
+      size_t field_id = FieldID(index);
+      int field_t_bits = sizeof(bit_set[0]) * 8;
```
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:731]: use this value instead of recalculating https://github.com/devosoft/Empirical/blob/62fccad1c411c3cbae4a64f9387dee01d1a66648/include/emp/bits/BitVector.hpp#L57

File: include/emp/bits/BitVector.hpp (hunk 2 above)
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:728]: I'd rephrase this as

Zero out bits lower than index

File: include/emp/bits/BitVector.hpp [hunk 9]
```diff
@@ -701,6 +700,72 @@ namespace emp {
...
+      if (index < num_bits)
+        bit_set[field_id] &= MaskLow<field_t>(index % field_t_bits);
+    }
+
+    /// Insert bits into vector, including middle using bitmagic
+    /// https://devolab.org/?p=2249
+    void Insert(size_t index, bool value=true, size_t num=1) {
+      Resize(num_bits + num);
+      // mask high and shift left: 101101 -> 1010000
+      thread_local BitVector a{};
```
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:743]: give a brief explanation of why we use thread_local (prevent re-allocation)

File: include/emp/bits/BitVector.hpp [hunk 10]
```diff
@@ -701,6 +700,72 @@ namespace emp {
...
+      a = *this;
+      if (index > 0)
+        a.Mask_High(index + num - 2);
+      a <<= num;
+      // mask low and | together 101101 -> 000101 ->1010101
+      thread_local BitVector b{};
+      b = *this;
+      b.Mask_Low(index);
+      a |= b;
+      // set specified bits -> 1010101 -> 101_101
```
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:753]: good explanation with the example!!!

File: include/emp/bits/BitVector.hpp [hunk 11]
```diff
@@ -701,6 +700,72 @@ namespace emp {
...
+      for (size_t i = index; i < index + num; i++) {
+        a.Set(i, value);
+      }
+      std::swap(a, *this);
+
+    }
+
+    /// Delete bits from vector, including middle
```
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:761]: what do you mean by "including middle?"

Comment (mmore500) [inline include/emp/bits/BitVector.hpp:761]: Be sure to write docstrings as complete sentence (including period)

File: include/emp/bits/BitVector.hpp [hunk 12]
```diff
@@ -701,6 +700,72 @@ namespace emp {
...
+    void Delete(size_t index, size_t num=1) {
+      for (size_t j = index; j < num_bits - num; j++) {
```
Comment (mmore500) [inline include/emp/bits/BitVector.hpp:763]: Add a TODO comment saying that we should consider a bitmagic approach here

Comment (mmore500) [inline include/emp/bits/BitVector.hpp:763]: where possible, try to name iterator variables descriptively (for example, bit in this case) or default to i instead of j

File: include/emp/tools/timing.hpp [hunk 13]
```diff
@@ -31,6 +31,15 @@
               << " ms" << std::endl;                                         \
     std::cout << "Result: " << emp_result << std::endl;                      \
   }
+/// A simple macro to time how long it takes for a function to complete.
```
Comment (mmore500) [inline include/emp/tools/timing.hpp:34]: add a line break above

File: tests/bits/BitVector.cpp [hunk 14]
```diff
@@ -8,10 +8,25 @@
...
 #include "emp/bits/BitVector.hpp"
 #include "emp/base/vector.hpp"
+#include "emp/tools/timing.hpp"
 
 #include <sstream>
 #include <map>
 #include <limits>
+#include <ratio>
+
+TEST_CASE("Benchmark BitVector Inserts", "[bits]"){
```
Comment (mmore500) [inline tests/bits/BitVector.cpp:18]: add a comment with approximately the timing you're observing on your computer

File: tests/bits/BitVector.cpp [hunk 15]
```diff
@@ -8,10 +8,25 @@
...
+	emp::BitVector bv(0);
+	bv.Insert(0, true, 4096);
+	std::cout << "Bit Magic Insert: ";
+	EMP_VOID_FUNCTION_TIMER([&bv](){
+		for ( size_t i = 1; i <= std::mega::num; ++i ) {
+			auto bv1 = bv;
+			bv1.Insert(1, false);
+			//bv.PopBack();
```
Comment (mmore500) [inline tests/bits/BitVector.cpp:26]: delete commented out code

File: tests/bits/BitVector.cpp [hunk 16]
```diff
@@ -259,6 +278,80 @@ TEST_CASE("Test BitVector", "[bits]")
 	REQUIRE(bv_f.count() == 1);
 	bv_f <<= 1;
 	REQUIRE(bv_f.none());
+
```
Comment (mmore500) [inline tests/bits/BitVector.cpp:281]: for new tests that we write, let's put each section in its own TEST_CASE

example TEST_CASE("Test Mask_High,Mask_low")

**Review by mmore500 — CHANGES_REQUESTED** [review]

**Comment by mmore500** [comment]
reminder to self: we should check codecov before merging this ini
